### ISSUE TITLE
Feature Request: Optimizer that can perform Sparse Updates

### ISSUE THREAD

**Issue by ethancaballero** [tensorflow/tensorflow#5326]
context:
https://www.reddit.com/r/MachineLearning/comments/5abcd4/r161009027_scaling_memoryaugmented_neural/d9fz62u/

Feature request to add optimizer that can perform sparse updates so that [1610.09027] "Scaling Memory-Augmented Neural Networks with Sparse Reads and Writes" from DeepMind can be implemented in Tensorflow.

@alrojo

**Comment by aselle** [comment]
We don't have any plans right now to do this. Thanks!

**Comment by alrojo** [comment]
@aselle given you support [sparse variable updates](https://www.tensorflow.org/versions/r0.11/api_docs/python/state_ops.html#sparse-variable-updates) and your optimizers have an [_apply_sparse](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/adam.py#L141) functionality.

Perhaps you could elaborate on the current state of sparse support for optimizers in TensorFlow and which challenges exists for implementing fully sparse support?

Thanks

**Comment by aselle** [comment]
I don't think optimizers are taking advantage of sparse updates yet. @alextp, do you have anything to add on the subject? It's an area of some interest of course, so thanks for bringing it up.

**Comment by alextp** [comment]
Currently optimizers in TF do sparse updates, as you pointed out, only if the variable access was made with a gather (as the gradient of gather is defined to be indexedslices which triggers sparse updates).

That said what this paper implemented is not sparse updates as part of the optimization process but sparse updates as part of unrolling an RNN (wherein each state has a reference to a very large memory tensor and does scatter_add on it). Currently tensorflow has no support for efficient sparse mutations on arbitrary tensors.

I see two ways of supporting this: adding a new type of mutable resource like tensorarray which allows for these operations, or adding a graph-rewriting pass which checks which sparse tensor mutations are valid and can be done with overwriting instead of copying.

**Comment by martinwicke** [comment]
This issue is quite old and hasn't had recent activity. If there's a well-scoped feature requests in here, please open a new issue. Thank you.
